core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -71,79 +75,226 @@ object TransactionMarkerChannelManager {
+
+  private val interBrokerListenerName: ListenerName = config.interBrokerListenerName
 
-  type WriteTxnMarkerCallback = Errors => Unit
+  // TODO: What is reasonable for this
+  private val brokerNotAliveBackoffMs = 10
+
+  private val txnMarkerSendThread: InterBrokerSendThread = {
+    new InterBrokerSendThread("TxnMarkerSenderThread-" + config.brokerId, networkClient, drainQueuedTransactionMarkers, time)
+  }
 
   def start(): Unit = {
-    interBrokerSendThread.start()
+    txnMarkerSendThread.start()
+    networkClient.wakeup()    // FIXME: is this really required?
   }
 
   def shutdown(): Unit = {
-    interBrokerSendThread.shutdown()
-    transactionMarkerChannel.close()
+    txnMarkerSendThread.shutdown()
+    markersQueuePerBroker.clear()
+  }
+
+  // visible for testing
+  private[transaction] def queueForBroker(brokerId: Int) = {
+    markersQueuePerBroker.get(brokerId)
+  }
+
+  // visible for testing
+  private[transaction] def senderThread = txnMarkerSendThread
+
+  private[transaction] def addMarkersForBroker(broker: Node, txnTopicPartition: Int, txnIdAndMarker: TxnIdAndMarkerEntry) {
+    val brokerId = broker.id
+
+    // we do not synchronize on the update of the broker node with the enqueuing,
+    // since even if there is a race condition we will just retry
+    val brokerRequestQueue = markersQueuePerBroker.getOrElseUpdate(brokerId, new TxnMarkerQueue(broker))
+    brokerRequestQueue.maybeUpdateNode(broker)
+    brokerRequestQueue.addMarkers(txnTopicPartition, txnIdAndMarker)
+
+    trace(s"Added marker ${txnIdAndMarker.txnMarkerEntry} for transactional id ${txnIdAndMarker.txnId} to destination broker $brokerId")
   }
 
+  private[transaction] def drainQueuedTransactionMarkers(): Iterable[RequestAndCompletionHandler] = {
+    markersQueuePerBroker.map { case (brokerId: Int, brokerRequestQueue: TxnMarkerQueue) =>
+      val txnIdAndMarkerEntries: java.util.List[TxnIdAndMarkerEntry] = new util.ArrayList[TxnIdAndMarkerEntry]()
+      brokerRequestQueue.forEachTxnTopicPartition { case (_, queue) =>
+        queue.drainTo(txnIdAndMarkerEntries)
+      }
+      (brokerRequestQueue.node, txnIdAndMarkerEntries)
+    }
+      .filter { case (_, entries) => !entries.isEmpty}
+      .map { case (node, entries) =>
+        val markersToSend: java.util.List[TxnMarkerEntry] = entries.asScala.map(_.txnMarkerEntry).asJava
+        val requestCompletionHandler = new TransactionMarkerRequestCompletionHandler(node.id, txnStateManager, this, entries)
+        RequestAndCompletionHandler(node, new WriteTxnMarkersRequest.Builder(markersToSend), requestCompletionHandler)
+      }
+  }
+
+  def addTxnMarkersToSend(transactionalId: String,
+                          coordinatorEpoch: Int,
+                          txnResult: TransactionResult,
+                          txnMetadata: TransactionMetadata,
+                          newMetadata: TransactionMetadataTransition): Unit = {
+
+    def appendToLogCallback(error: Errors): Unit = {
```

Comment:
Hmm that is right, but I feel we will probably need to add other possible error codes as we as fixing various error cases under integration tests, so I'd rather keep it as is for now. If after exactly-once has been quite stable and we still do not have any other error code we can remove it. 

But let mw add a TODO for now in case we forgot..